File: frontend/src/types/graph.ts
/**
 * Ideograph - Dependency Graph for Startup Idea Evaluation
 * Core Type Definitions
 */

// =============================================================================
// NODE TYPES
// =============================================================================

/**
 * The five fundamental node types in the evaluation graph
 */
export type NodeType = 'framework' | 'claim' | 'fact' | 'evidence' | 'startup_meta';

/**
 * The three PMF paths that determine startup strategy
 */
export type PmfPath = 'hair_on_fire' | 'hard_fact' | 'future_vision';

/**
 * Configuration for each PMF path
 */
export interface PmfPathConfig {
  id: PmfPath;
  name: string;
  icon: string;
  shortDescription: string;
  enemy: string;
  job: string;
  speedMatters: string;
  benchmark: string;
}

// =============================================================================
// CONFIDENCE BANDS
// =============================================================================

/**
 * Confidence band levels - honest uncertainty instead of false precision
 */
export type ConfidenceBandLevel = 'high' | 'medium' | 'low' | 'assumption';

export interface ConfidenceBand {
  level: ConfidenceBandLevel;
  label: string;
  description: string;
  color: string;
}

export const CONFIDENCE_BANDS: Record<ConfidenceBandLevel, ConfidenceBand> = {
  high: {
    level: 'high',
    label: 'Strong evidence',
    description: 'Multiple credible sources agree. Still could be wrong.',
    color: '#007354', // Sequoia green
  },
  medium: {
    level: 'medium',
    label: 'Some evidence',
    description: 'Reasonable basis but needs more validation.',
    color: '#E5A826', // Warning yellow
  },
  low: {
    level: 'low',
    label: 'Weak evidence',
    description: 'Limited data. Treat as hypothesis to test.',
    color: '#FB923C', // Orange
  },
  assumption: {
    level: 'assumption',
    label: 'Assumption',
    description: 'No real evidence. Needs validation before building on it.',
    color: '#D4442E', // Red
  },
};

/**
 * Get confidence band from internal score
 * Internal scores are used for calculation only - never shown to users
 */
export function getConfidenceBandFromScore(score: number): ConfidenceBand {
  const normalized = score / 100;
  if (normalized >= 0.75) return CONFIDENCE_BANDS.high;
  if (normalized >= 0.55) return CONFIDENCE_BANDS.medium;
  if (normalized >= 0.35) return CONFIDENCE_BANDS.low;
  return CONFIDENCE_BANDS.assumption;
}

/**
 * Get edge color based on confidence band
 */
export function getEdgeColorFromConfidence(confidence: number): string {
  return getConfidenceBandFromScore(confidence).color;
}

export const PMF_PATH_CONFIGS: Record<PmfPath, PmfPathConfig> = {
  hair_on_fire: {
    id: 'hair_on_fire',
    name: 'Hair on Fire',
    icon: '🔥',
    shortDescription: 'Customer is actively searching for solutions RIGHT NOW',
    enemy: 'Competitors',
    job: 'Outexecute',
    speedMatters: 'Everything',
    benchmark: 'Wiz ($0→$100M in 18mo)',
  },
  hard_fact: {
    id: 'hard_fact',
    name: 'Hard Fact',
    icon: '📊',
    shortDescription: 'Customer has accepted the problem as "just how things are"',
    enemy: 'Inertia',
    job: 'Educate',
    speedMatters: 'Less critical',
    benchmark: 'HubSpot (coined "inbound")',
  },
  future_vision: {
    id: 'future_vision',
    name: 'Future Vision',
    icon: '🔮',
    shortDescription: "Customer doesn't know they have the problem",
    enemy: 'Disbelief',
    job: 'Survive',
    speedMatters: 'Decades',
    benchmark: 'Nvidia (25 years to AI)',
  },
};

/**
 * Evidence can be either supporting (validates) or contradicting (invalidates)
 */
export type EvidenceValence = 'supporting' | 'contradicting';

/**
 * Visual configuration for each node type
 */
export interface NodeVisualConfig {
  color: string;
  shape: 'square' | 'circle';
}

export const NODE_VISUALS: Record<NodeType, NodeVisualConfig> = {
  framework: { color: '#808080', shape: 'square' },  // Gray, square
  claim: { color: '#F5C542', shape: 'circle' },      // Yellow, circle
  fact: { color: '#4A90D9', shape: 'circle' },       // Blue, circle
  evidence: { color: '#22C55E', shape: 'circle' },   // Default green (can be red if contradicting)
  startup_meta: { color: '#007354', shape: 'square' }, // Sequoia green, square (hidden node)
};

export const EVIDENCE_COLORS: Record<EvidenceValence, string> = {
  supporting: '#22C55E',    // Green
  contradicting: '#EF4444', // Red
};

// =============================================================================
// VERSION HISTORY
// =============================================================================

/**
 * A single version entry tracking content changes
 */
export interface NodeVersion {
  id: string;
  content: string;
  timestamp: Date;
  changeNote?: string;
}

// =============================================================================
// BASE NODE & SPECIALIZED NODES
// =============================================================================

/**
 * Base properties shared by all nodes
 */
export interface BaseNode {
  id: string;
  type: NodeType;
  title: string;
  content: string;
  confidence: number; // 0-100
  createdAt: Date;
  updatedAt: Date;
  versions: NodeVersion[];
  metadata?: Record<string, unknown>;
}

/**
 * Position hint for visual layout
 */
export interface PositionHint {
  x: number;
  y: number;
  level: number; // 0 = root, 1 = category, 2 = leaf
}

/**
 * Framework node - fixed structure from evaluation skeleton
 * Represents structural elements that organize the evaluation
 * These nodes are IMMUTABLE - users cannot delete or rename them
 */
export interface FrameworkNode extends BaseNode {
  type: 'framework';
  /** Framework nodes are part of a predefined structure */
  frameworkCategory?: FrameworkCategory;
  /** Description explaining what this node is asking the founder to address */
  description: string;
  /** Position hint for visual layout */
  position?: PositionHint;
  /** Whether this is a skeleton node (immutable) */
  isSkeleton: boolean;
  /** Parent framework node ID (for skeleton hierarchy) */
  parentId?: string;
}

/**
 * The 8 framework categories
 */
export type FrameworkCategory = 
  | 'root'
  | 'market'
  | 'problem'
  | 'solution'
  | 'timing'
  | 'founder'
  | 'competition'
  | 'business_model'
  | 'gtm';

/**
 * Claim node - something the founder believes, unvalidated
 * These need to be tested against evidence
 */
export interface ClaimNode extends BaseNode {
  type: 'claim';
  /** Who made this claim */
  source?: string;
  /** Has this claim been tested yet? */
  tested: boolean;
}

/**
 * Fact node - objectively true information
 * These are verified truths that inform the evaluation
 */
export interface FactNode extends BaseNode {
  type: 'fact';
  /** Source of this fact (research, data, etc.) */
  source?: string;
  /** URL or reference to verification */
  verificationUrl?: string;
}

/**
 * Evidence node - proof from real world
 * Can either support or contradict other nodes
 */
export interface EvidenceNode extends BaseNode {
  type: 'evidence';
  /** Supporting evidence validates, contradicting evidence invalidates */
  valence: EvidenceValence;
  /** Where this evidence came from */
  source: string;
  /** Date when this evidence was gathered */
  gatheredAt?: Date;
}

/**
 * Startup Meta node - stores startup-wide metadata
 * Only one should exist per graph
 */
export interface StartupMetaNode extends BaseNode {
  type: 'startup_meta';
  /** The detected PMF path */
  pmfPath?: PmfPath;
  /** Confidence in the path detection (0-100) */
  pmfPathConfidence: number;
  /** When the path was detected/set */
  pmfPathDetectedAt?: Date;
  /** Optional startup name */
  startupName?: string;
}

/**
 * Union type of all possible nodes
 */
export type GraphNode = FrameworkNode | ClaimNode | FactNode | EvidenceNode | StartupMetaNode;

// =============================================================================
// EDGE TYPES
// =============================================================================

/**
 * The relationship types between nodes
 */
export type EdgeType = 
  | 'requires'     // Parent needs child to be true
  | 'supports'     // Child makes parent more likely
  | 'contradicts'  // Child makes parent less likely
  | 'depends_on'   // If child fails, parent fails
  | 'informs'      // Child provides context for parent
  | 'blocks';      // Child prevents parent

/**
 * Visual and semantic configuration for edge types
 */
export interface EdgeTypeConfig {
  color: string;
  label: string;
  description: string;
  /** Does this edge type imply a negative relationship? */
  isNegative: boolean;
}

export const EDGE_CONFIGS: Record<EdgeType, EdgeTypeConfig> = {
  requires: {
    color: '#6B7280',
    label: 'requires',
    description: 'Parent needs child to be true',
    isNegative: false,
  },
  supports: {
    color: '#22C55E',
    label: 'supports',
    description: 'Child makes parent more likely',
    isNegative: false,
  },
  contradicts: {
    color: '#EF4444',
    label: 'contradicts',
    description: 'Child makes parent less likely',
    isNegative: true,
  },
  depends_on: {
    color: '#F59E0B',
    label: 'depends on',
    description: 'If child fails, parent fails',
    isNegative: false,
  },
  informs: {
    color: '#3B82F6',
    label: 'informs',
    description: 'Child provides context for parent',
    isNegative: false,
  },
  blocks: {
    color: '#DC2626',
    label: 'blocks',
    description: 'Child prevents parent',
    isNegative: true,
  },
};

// =============================================================================
// EDGE DEFINITION
// =============================================================================

/**
 * An edge connecting two nodes in the graph
 * Direction: from (child/source) -> to (parent/target)
 */
export interface GraphEdge {
  id: string;
  type: EdgeType;
  from: string; // Source node ID (child)
  to: string;   // Target node ID (parent)
  /** Optional weight/strength of the relationship (0-100) */
  weight?: number;
  /** Optional note explaining this relationship */
  note?: string;
  createdAt: Date;
}

// =============================================================================
// GRAPH STRUCTURE
// =============================================================================

/**
 * The complete graph state
 */
export interface GraphState {
  nodes: Map<string, GraphNode>;
  edges: Map<string, GraphEdge>;
  /** Index for fast lookup of edges by node */
  nodeEdgeIndex: {
    outgoing: Map<string, Set<string>>; // nodeId -> set of edge IDs going out
    incoming: Map<string, Set<string>>; // nodeId -> set of edge IDs coming in
  };
}

// =============================================================================
// HELPER TYPES
// =============================================================================

/**
 * Input type for creating a new node (without auto-generated fields)
 */
export type CreateNodeInput<T extends GraphNode = GraphNode> = Omit<
  T,
  'id' | 'createdAt' | 'updatedAt' | 'versions'
> & {
  id?: string; // Optional custom ID
};

/**
 * Input type for creating a new edge
 */
export type CreateEdgeInput = Omit<GraphEdge, 'id' | 'createdAt'> & {
  id?: string;
};

/**
 * Input type for updating a node
 */
export type UpdateNodeInput = Partial<Omit<GraphNode, 'id' | 'type' | 'createdAt' | 'versions'>> & {
  changeNote?: string;
};

/**
 * Query result for connected nodes
 */
export interface ConnectedNodesResult {
  node: GraphNode;
  edge: GraphEdge;
  direction: 'parent' | 'child';
}

File: backend/app/models/graph.py
"""
Graph Data Models
Pydantic models for the Ideograph dependency graph
"""

from datetime import datetime
from enum import Enum
from typing import Optional, Union, List, Dict, Any
from pydantic import BaseModel, Field
import uuid


# =============================================================================
# ENUMS
# =============================================================================

class NodeType(str, Enum):
    FRAMEWORK = "framework"
    CLAIM = "claim"
    FACT = "fact"
    EVIDENCE = "evidence"
    STARTUP_META = "startup_meta"


class PmfPath(str, Enum):
    """The three fundamental PMF paths that determine startup strategy"""
    HAIR_ON_FIRE = "hair_on_fire"      # Customer actively searching for solutions
    HARD_FACT = "hard_fact"            # Customer has accepted the problem as "just how things are"
    FUTURE_VISION = "future_vision"    # Customer doesn't know they have the problem


class EdgeType(str, Enum):
    REQUIRES = "requires"
    SUPPORTS = "supports"
    CONTRADICTS = "contradicts"
    DEPENDS_ON = "depends_on"
    INFORMS = "informs"
    BLOCKS = "blocks"


class FrameworkCategory(str, Enum):
    ROOT = "root"
    MARKET = "market"
    PROBLEM = "problem"
    SOLUTION = "solution"
    TIMING = "timing"
    FOUNDER = "founder"
    COMPETITION = "competition"
    BUSINESS_MODEL = "business_model"
    GTM = "gtm"


class EvidenceValence(str, Enum):
    SUPPORTING = "supporting"
    CONTRADICTING = "contradicting"


# =============================================================================
# POSITION
# =============================================================================

class PositionHint(BaseModel):
    x: float
    y: float
    level: int  # 0 = root, 1 = category, 2 = leaf


# =============================================================================
# BASE NODE
# =============================================================================

class BaseNode(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    type: NodeType
    title: str
    content: str = ""
    confidence: int = Field(default=50, ge=0, le=100)
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    metadata: Optional[Dict[str, Any]] = None

    class Config:
        use_enum_values = True


# =============================================================================
# SPECIALIZED NODES
# =============================================================================

class FrameworkNode(BaseNode):
    type: NodeType = NodeType.FRAMEWORK
    framework_category: Optional[FrameworkCategory] = None
    description: str = ""
    position: Optional[PositionHint] = None
    is_skeleton: bool = True
    parent_id: Optional[str] = None


class ClaimNode(BaseNode):
    type: NodeType = NodeType.CLAIM
    source: Optional[str] = None
    tested: bool = False


class FactNode(BaseNode):
    type: NodeType = NodeType.FACT
    source: Optional[str] = None
    verification_url: Optional[str] = None


class EvidenceNode(BaseNode):
    type: NodeType = NodeType.EVIDENCE
    valence: EvidenceValence = EvidenceValence.SUPPORTING
    source: str = ""
    gathered_at: Optional[datetime] = None


class StartupMetaNode(BaseNode):
    """
    Special node that stores startup-level metadata.
    Only one of these should exist per graph.
    Stores PMF path and other startup-wide settings.
    """
    type: NodeType = NodeType.STARTUP_META
    pmf_path: Optional[PmfPath] = None
    pmf_path_confidence: int = Field(default=100, ge=0, le=100)
    pmf_path_detected_at: Optional[datetime] = None
    startup_name: Optional[str] = None


# Union type for all nodes
GraphNode = Union[FrameworkNode, ClaimNode, FactNode, EvidenceNode, StartupMetaNode]


# =============================================================================
# EDGE
# =============================================================================

class GraphEdge(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    type: EdgeType
    from_node: str = Field(alias="from")
    to_node: str = Field(alias="to")
    note: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.now)

    class Config:
        use_enum_values = True
        populate_by_name = True


# =============================================================================
# INPUT MODELS
# =============================================================================

class CreateNodeInput(BaseModel):
    id: Optional[str] = None
    type: NodeType
    title: str
    content: str = ""
    confidence: int = Field(default=50, ge=0, le=100)
    
    # Framework-specific
    framework_category: Optional[FrameworkCategory] = None
    description: Optional[str] = None
    position: Optional[PositionHint] = None
    is_skeleton: Optional[bool] = None
    parent_id: Optional[str] = None
    
    # Claim-specific
    source: Optional[str] = None
    tested: Optional[bool] = None
    
    # Evidence-specific
    valence: Optional[EvidenceValence] = None

    class Config:
        use_enum_values = True


class CreateEdgeInput(BaseModel):
    id: Optional[str] = None
    type: EdgeType
    from_node: str = Field(alias="from")
    to_node: str = Field(alias="to")
    note: Optional[str] = None

    class Config:
        use_enum_values = True
        populate_by_name = True


class UpdateNodeInput(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    confidence: Optional[int] = Field(default=None, ge=0, le=100)
    change_note: Optional[str] = None


# =============================================================================
# RESPONSE MODELS
# =============================================================================

class GraphState(BaseModel):
    nodes: List[Dict[str, Any]]
    edges: List[Dict[str, Any]]


class NodeResponse(BaseModel):
    success: bool
    node: Optional[Dict[str, Any]] = None
    message: Optional[str] = None


class EdgeResponse(BaseModel):
    success: bool
    edge: Optional[Dict[str, Any]] = None
    message: Optional[str] = None


class ProgressResponse(BaseModel):
    answered: int
    total: int
    percentage: int


# =============================================================================
# GAP ANALYSIS MODELS
# =============================================================================

class GapResponse(BaseModel):
    """A single gap in framework coverage"""
    framework_id: str
    question: str
    why_it_matters: str
    criticality: str  # killer, high, medium, low
    founder_trap: str
    probing_questions: List[str]
    coverage_score: float
    connected_nodes_count: int


class PriorityAction(BaseModel):
    """A prioritized action to address a gap"""
    framework_id: str
    action: str
    criticality: str
    why: str


class GapSummary(BaseModel):
    """Summary of all gaps in the knowledge graph"""
    total_gaps: int
    killer_gaps: int
    high_gaps: int
    medium_gaps: int
    low_gaps: int
    most_critical: Optional[GapResponse] = None
    priority_actions: List[PriorityAction]
    overall_readiness: str  # critical_gaps, needs_work, in_progress, good, complete
    completion_percentage: float


class GapAnalysisResponse(BaseModel):
    """Full gap analysis response"""
    success: bool
    summary: GapSummary
    gaps: List[GapResponse]
    category_breakdown: Optional[Dict[str, Any]] = None


class CategoryGapsResponse(BaseModel):
    """Gap analysis for a specific category"""
    category: str
    total_questions: int
    gaps_count: int
    completion_percentage: float
    gaps: List[GapResponse]
